torchtitan/experiments/deepseek_v3/moe_kernels.py

```diff
@@ -0,0 +1,307 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Tuple
+
+import torch
+import triton
+import triton.language as tl
+
+
+# parallelized kernel
+@triton.jit
+def _fill_indices_kernel(
+    tokens_per_expert_group_ptr,
+    start_index_values_ptr,
+    write_offsets_ptr,
+    output_ptr,
+    experts_per_rank,
+    num_ranks,
+    BLOCK_SIZE: tl.constexpr,  # Number of threads per block
+):
+    pid = tl.program_id(axis=0)
+
+    # each program handles one expert
+    expert_id = pid
+
+    # only process if valid expert
+    if expert_id < experts_per_rank:
```

Comment:
After adding the cap, we may need modulo and for loop around expert_id

Reply:
add a for loop around this based on start and end calcs for total num experts.
also added more unit tests to cover this scenario (i.e. verify results work assuming looping).